### ISSUE TITLE
Update to Unicode 15.1

### ISSUE THREAD

**Issue by krlmlr** [patperry/utf8lite#2]
@patperry: what do we need to do to update the Unicode tables to the most recent version? Is it worth adding GitHub Actions automation?

**Comment by patperry** [comment]
This is a little tricky, as Unicode has changed the [Grapheme Break Rules](https://unicode.org/reports/tr29/). Rule GB9c is new.

You can change the Makefile, set `UNICODE = http://www.unicode.org/Public/15.1.0` then run make. This will update all the tables. However, when you run `make check` you will see that some of the grapheme boundary tests fail.

I'm not sure when/if I'll be able to look into this. Are you or someone else able? You'd need to update [graphscan.c](https://github.com/patperry/utf8lite/blob/master/src/graphscan.c)

**Comment by patperry** [comment]
I just updated to [15.0.0](https://github.com/patperry/utf8lite/commit/644584e27c9418c8209b498ade9f809e56b4ab51).

**Comment by krlmlr** [comment]
So the change to 15.1.0 still needs a code update?

CC @gaborcsardi.

**Comment by gaborcsardi** [comment]
FWIW I didn't see any check errors with 15.1.0. In a clean build:
```
❯ make check
tests/check_charwidth
Running suite(s): charwidth
100%: Checks: 2, Failures: 0, Errors: 0
tests/check_graphscan
Running suite(s): graphscan
100%: Checks: 7, Failures: 0, Errors: 0
tests/check_render
Running suite(s): render
100%: Checks: 35, Failures: 0, Errors: 0
tests/check_text
Running suite(s): text
100%: Checks: 75, Failures: 0, Errors: 0
tests/check_textmap
Running suite(s): textmap
100%: Checks: 10, Failures: 0, Errors: 0
tests/check_unicode
Running suite(s): unicode
100%: Checks: 14, Failures: 0, Errors: 0
tests/check_wordscan
Running suite(s): wordscan
100%: Checks: 4, Failures: 0, Errors: 0
```

**Comment by patperry** [comment]
@gaborcsardi you need to delete the `data` directory before running `make` and `make check`.

I see the following:
```
laptop~/repos/utf8lite(master|✚15…) % make check
tests/check_charwidth
Running suite(s): charwidth
100%: Checks: 2, Failures: 0, Errors: 0
tests/check_graphscan
Running suite(s): graphscan
71%: Checks: 7, Failures: 2, Errors: 0
tests/check_graphscan.c:309:F:Unicode GraphemeBreakTest.txt:test_unicode_forward:0: Assertion 'scan.current.text.ptr + ((&scan.current.text)->attr & ((size_t)18446744073709551615UL >> 1)) == test->break_end[j]' failed
tests/check_graphscan.c:338:F:Unicode GraphemeBreakTest.txt:test_unicode_backward:0: Assertion 'scan.current.text.ptr == test->break_begin[j]' failed
make: *** [tests/check_graphscan.test] Error 1
```

**Comment by gaborcsardi** [comment]
Right, that makes sense, I only did `git clean -fdx .`...

**Comment by patperry** [comment]
I started a branch for this, still in progress: https://github.com/patperry/utf8lite/tree/unicode-15.1

**Comment by patperry** [comment]
This is implemented now, merged into master in b2d8f68 .

**Comment by krlmlr** [comment]
Brilliant! I wonder if, medium term, the utf8 R package could use cli under the hood. Need to check if anything's missing.

In the short term, it might be easier to update the bundled library and release.

**Comment by gaborcsardi** [comment]
I think the utf8 package is great, in that it is small and standalone. Ideally, cli would depend on utf8, not the other way around, but we want cli to be dependency free, and also to have access to the UTF-8 functions from C code.
